bugbug/vectordb.py

```diff
@@ -0,0 +1,125 @@
+# -*- coding: utf-8 -*-
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import re
+from abc import ABC, abstractmethod
+from dataclasses import asdict, dataclass
+from itertools import islice
+from typing import Iterable
+
+from langchain_openai import OpenAIEmbeddings
+from qdrant_client import QdrantClient
+from qdrant_client.http.exceptions import UnexpectedResponse
+from qdrant_client.models import Distance, PointStruct, VectorParams
+from unidiff import Hunk
+
+from bugbug.tools.code_review import InlineComment
+
+
+# TODO: Replace this with `from itertools import batched` once we move to
+# Python 3.12. The function below was copied from Itertools recipes.
+# Source: https://docs.python.org/3.11/library/itertools.html#itertools-recipes
+def batched(iterable, n):
+    """Batch data into tuples of length n. The last batch may be shorter."""
+    # batched('ABCDEFG', 3) --> ABC DEF G
+    if n < 1:
+        raise ValueError("n must be at least one")
+    it = iter(iterable)
+    while batch := tuple(islice(it, n)):
+        yield batch
+
+
+@dataclass
+class VectorPoint:
+    id: int
+    vector: list[float]
+    payload: dict
+
+
+class VectorDB(ABC):
+    """Abstract class for a vector database.
+
+    You can implement this class to support different vector databases.
+    """
+
+    @abstractmethod
+    def setup(self):
+        ...
+
+    @abstractmethod
+    def upsert(self, points: Iterable[VectorPoint]):
+        ...
+
+    @abstractmethod
+    def search(self, query: list[float]):
+        ...
+
+
+class QdrantVectorDB(VectorDB):
+    def __init__(
+        self, location: str, api_key: str, collection_name: str, *args, **kwargs
+    ):
+        super().__init__(*args, **kwargs)
+        self.collection_name = collection_name
+        self.client = QdrantClient(location=location, api_key=api_key)
+
+    def setup(self):
+        config = VectorParams(
+            size=3072,
```

Comment:
What is `size`? Should it be configurable through a param?

Reply:
> What is `size`? 

It is the size of the vector. So, it is 3072 here to match the size of the vector generated by the `text-embedding-3-large` model.

> Should it be configurable through a param?

At this early stage, I prefer to delay exposing configurations until they are needed. This will give us the flexibility to move things around without breaking the API.